storage/integration_test.go

```diff
@@ -2799,197 +2797,158 @@ func TestIntegration_RequesterPays(t *testing.T) {
-		}
-		// user: an Owner on the containing project
-		// userProject: containing project
-		// result: success, by the same rule as above; userProject is unnecessary but allowed.
-		if err := f(b1a.UserProject(projID)); err != nil {
-			t.Errorf("%s: got %v, want nil", msg, err)
-		}
-		// user: not an Owner on the containing project
-		// userProject: absent
-		// result: failure, by the standard requester-pays rule
-		err := f(b2a)
-		if got, want := errCode(err), wantErrorCode; got != want {
-			t.Errorf("%s: got error %v with code %d, want code %d\n"+
-				"confirm that %s is NOT an Owner on %s",
-				msg, err, got, want, otherUser, projID)
-		}
-		// user: not an Owner on the containing project
-		// userProject: not the containing one, but user has Editor role on it
-		// result: success, by the standard requester-pays rule
-		if err := f(b2a.UserProject(otherProjID)); err != nil {
-			t.Errorf("%s: got %v, want nil\n"+
-				"confirm that %s is an Editor on %s and that that project has billing enabled",
-				msg, err, otherUser, otherProjID)
-		}
-		// user: not an Owner on the containing project
-		// userProject: the containing one, on which the user does NOT have Editor permission.
-		// result: failure
-		err = f(b2.UserProject(noPermissionsProjID))
-		if got, want := errCode(err), 403; got != want {
-			t.Errorf("%s: got error %v, want code %d\n"+
-				"confirm that %s is NOT an Editor on %s",
-				msg, err, want, otherUser, noPermissionsProjID)
-		}
-	}
-
-	// Getting its attributes requires a user project.
-	var attrs *BucketAttrs
-	call("Bucket attrs", func(b *BucketHandle) error {
-		a, err := b.Attrs(ctx)
-		if a != nil {
-			attrs = a
-		}
-		return err
-	})
-	if attrs != nil {
-		if got, want := attrs.RequesterPays, true; got != want {
-			t.Fatalf("attr.RequesterPays = %t, want %t", got, want)
-		}
-	}
-	// Object operations.
-	call("write object", func(b *BucketHandle) error {
-		return writeObject(ctx, b.Object(obj1), "text/plain", []byte("hello"))
-	})
+	defer h.mustDeleteBucket(requesterPaysBucket)
 
-	call("read object", func(b *BucketHandle) error {
-		_, err := readObject(ctx, b.Object(obj1))
-		return err
-	})
-	call("object attrs", func(b *BucketHandle) error {
-		_, err := b.Object(obj1).Attrs(ctx)
-		return err
-	})
-	call("update object", func(b *BucketHandle) error {
-		_, err := b.Object(obj1).Update(ctx, ObjectAttrsToUpdate{ContentLanguage: "en"})
-		return err
-	})
+	for _, test := range []struct {
+		desc          string
+		client        *Client
+		userProject   *string // to set on bucket, nil if it should not be set
+		expectSuccess bool
+		wantErrorCode int
+		errorMessage  string
+	}{
+		{
+			desc:          "user is Owner on the project that owns the bucket",
+			client:        mainUserClient,
+			userProject:   nil,
+			expectSuccess: true, // by the rule permitting access by owners of the containing bucket
+			errorMessage:  fmt.Sprintf("confirm that %s is an Owner on %s", mainUserEmail, mainProjectID),
```

Comment:
Checking the specific error messages is probably too brittle. We can check status code and (potentially) reason but the full message is subject to change. Also, why is there an error message on cases which are expected to succeed?

Reply:
So this error message is the one we return in case of error, not an error we are checking for.
For each test case we want specific conditions to hold (for example, mainUser should be an Owner), otherwise the test case won't pass, so we add that condition (defined in `errorMessage`) to the output to help debugging in case of test failure. 
These were the messages from the original test. I can see where this is confusing though - do you have any suggestions to improve how we do this?